docs/snapshot-versioning.md

```diff
@@ -0,0 +1,96 @@
+
+![Version graph](
+images/version_graph.png?raw=true
+"Version graph")
+
+## Overview
+Firecracker persists the microvm state as 2 separate objects:
+ - a  **guest memory** file
+ - a **vmstate** file
+ 
+*The block devices attached to the microvm are not considered part of the state and need to be managed separately.*
+
+### Guest memory
+The guest memory file is agnostic to Firecracker's version tolerant save/restore as it's contents are fully owned by the guest OS. The microVM memory is by choice saved as a dump of each page or as a sparse file of guest modified pages only. 
+### Vmstate
+Stores the internal state of the VMM (device emulation , KVM and vCPUs) with 2 expcetions - serial emulation and vsock backend. 
+
+While we continuously improve and extend Firecracker's features by adding new options, devices or enhancements, the vmstate file changes both structurally and semantically with each new release. The state file includes versioning information and each Firecracker release implements distinct save/load logic for the supported version space. 
+
+## Vmstate file format
+
+###  A closer look
+A vm state file is further split in 4 different fields:
+
+| field | bits| description |
+|----|----|----|
+| magic_id | 64 | firecracker snapshot, architecture (x86_64/aarch64) and storage version.
+| version  | 64 | the snapshot version number internally 1:1 mapped to specific Firecracker version.
+| state | N*8 | Bincode blob containing the vm state.
+ | crc| 64 | Optional CRC64 sum of magic_id, version and state.
+
+Note: the last 16 bits of `magic_id` encode the storage version which specifies the encoding used for the `version` and `state` fields. The current implementation sets this field to 1 which identifies it as a [Serde bincode](https://github.com/servo/bincode) compatible encoder/decoder.
+
+### Version tolerant ser/de
+Firecracker reads and writes the `state` blob of the snapshot by using per version, separate serialization and deserialization logic. This logic is mostly autogenerated by a Rust procedural macro based on `struct`, `enum` and `union` annotations. Basically, one can say that these structures support versioning. The versioning logic is generated by parsing a structure's history log (encoded using Rust annotations) and emitting rust code.
+
+Versioned serialization and deserialization is divided in 2 translation layers:
+ - structural translator
+ - semantic translator
+
+The structural translator implements the logic to deserialize a structure field by field from any of it's previous versions and recreate the appropriate Rust POD structure at the latest version. It also implements the reverse: given a Rust POD structure it will serialize only the fields present at a specific target version.
```

Comment:
```suggestion
The structural translator implements the logic to deserialize a structure field by field from any of its previous versions and recreate the appropriate Rust POD structure at the latest version. It also implements the reverse: given a Rust POD structure it will serialize only the fields present at a specific target version.
```